File: python3/Trees/maximum_path_sum.py
from ds import TreeNode

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
"""


max_sum = float('-inf')

def max_path_sum(root: TreeNode) -> int:
    global max_sum
    max_path_sum_helper(root)
    return max_sum

def max_path_sum_helper(node: TreeNode) -> int:
    global max_sum
    # Base case: null nodes have no path sum.
    if not node:
        return 0
    # Collect the maximum gain we can attain from the left and right 
    # subtrees, setting them to 0 if they're negative.
    left_sum = max(max_path_sum_helper(node.left), 0)
    right_sum = max(max_path_sum_helper(node.right), 0)
    # Update the overall maximum path sum if the current path sum is 
    # larger.
    max_sum = max(max_sum, node.val + left_sum + right_sum)
    # Return the maximum sum of a single, continuous path with the 
    # current node as an endpoint.
    return node.val + max(left_sum, right_sum)

File: cpp/Trees/maximum_path_sum.cpp
#include <algorithm>
#include <limits>
#include "ds/TreeNode.h"
using ds::TreeNode;

/**
 * Definition of TreeNode:
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

int maxPathSum(TreeNode* root) {
    int maxSum = std::numeric_limits<int>::min();
    maxPathSumHelper(root, maxSum);
    return maxSum;
}

int maxPathSumHelper(TreeNode* node, int& maxSum) {
    // Base case: null nodes have no path sum.
    if (!node)
        return 0;
    // Collect the maximum gain we can attain from the left and right
    // subtrees, setting them to 0 if they're negative.
    int leftSum = std::max(maxPathSumHelper(node->left, maxSum), 0);
    int rightSum = std::max(maxPathSumHelper(node->right, maxSum), 0);
    // Update the overall maximum path sum if the current path sum is 
    // larger.
    maxSum = std::max(maxSum, node->val + leftSum + rightSum);
    // Return the maximum sum of a single, continuous path with the 
    // current node as an endpoint.
    return node->val + std::max(leftSum, rightSum);
}